atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/ValidatingTransactionScopedCache.java

```diff
@@ -0,0 +1,110 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.api.cache;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.atlasdb.futures.AtlasFutures;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.transaction.api.TransactionFailedNonRetriableException;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.logsafe.UnsafeArg;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class ValidatingTransactionScopedCache implements TransactionScopedCache {
+    private static final Logger log = LoggerFactory.getLogger(ValidatingTransactionScopedCache.class);
+
+    private final TransactionScopedCache delegate;
+    private final Supplier<Boolean> validationSupplier;
+
+    @VisibleForTesting
+    ValidatingTransactionScopedCache(TransactionScopedCache delegate, Supplier<Boolean> validationSupplier) {
+        this.delegate = delegate;
+        this.validationSupplier = validationSupplier;
+    }
+
+    static ValidatingTransactionScopedCache create(TransactionScopedCache delegate, double validationProbability) {
+        return new ValidatingTransactionScopedCache(
+                delegate, () -> ThreadLocalRandom.current().nextDouble() < validationProbability);
+    }
+
+    @Override
+    public void write(TableReference tableReference, Map<Cell, byte[]> values) {
+        delegate.write(tableReference, values);
+    }
+
+    @Override
+    public void delete(TableReference tableReference, Set<Cell> cells) {
+        delegate.delete(tableReference, cells);
+    }
+
+    @Override
+    public Map<Cell, byte[]> get(
+            TableReference tableReference,
+            Set<Cell> cell,
+            BiFunction<TableReference, Set<Cell>, ListenableFuture<Map<Cell, byte[]>>> valueLoader) {
+        if (validationSupplier.get()) {
+            Map<Cell, byte[]> remoteReads = AtlasFutures.getUnchecked(valueLoader.apply(tableReference, cell));
+            Map<Cell, byte[]> cacheReads =
+                    delegate.get(tableReference, cell, (table, cells) -> getCells(remoteReads, cells));
+
+            validateCacheReads(tableReference, remoteReads, cacheReads);
```

Comment:
Ah I wanted to complain that throwing is not ideal, but I see in the comment that that is to come -- is the idea to then disable caching for all transactions until a bounce happens?

Reply:
Yes